keps/sig-node/20191014-configurable-system-reserved.md

```diff
@@ -0,0 +1,183 @@
+If Kubernetes would also support the same feature, operators would be able to 1: isolate a common workload CPU pool from the operating system and host processes (see user story 1), and 2: manually divide this pool between the different infrastructures however they see fit.
+
+#### User Story 3
+As a TelCo edge cloud operator, I would like the CPUs of my latency sensitive applications to be managed by "out-of-tree" CPU managers
+
+Basically the culmination of the previous two use-cases. The CPU manager can be called e.g. Nova if the latency sensitive application is running in VMs, in which case use-case 3 equals use-case 2.
+But it could be also called CMK, or CPU-Pooler in case the latency sensitive application is running on top of Kubernetes.
+Kubelet's current CPU Manager is not used in neither of the cases.
+Note: the user story is taken -with some paraphrasing- from real life RFPs
+
+### Implementation Details/Notes/Constraints
+
+Kubernetes already contains code to remove a couple of CPU cores from the cpusets ("pools") it manages.
+The three possible enhancements to be considered are the following:
+- how to provide the possibility to optionally configure an exact cpuset as system-reserved
+- how the exclusion should inter-work with other, similar features
+- how to extend the scope of the exclusion to non-static policy use-cases too
+
+#### Configuration
+To avoid backward incompatible configuration API changes the KEP proposes to introduce a new configuration flag, called "reserved-cpus".
+This setting should be a Node-level / Kubelet option, and should be configurable independently of all other flags.
+
+This means inter-working with the existing CPU reservation mechanism for system processes has to be fully fleshed out, most importantly, with "system-reserved=cpu".
+Proposal is to let both of the flags live at the same time. When both are set "reserved-cpus" simply takes priority.
+
+#### Inter-working with other features
+As Kubelet already has multiple other features related to lowering a Node's allocatable CPU capacity, it shall be defined how the additional Node capacity reduction due to the newly introduced flag should work exactly.
+The idea is that whenever the Allocatable capacity of a Node is being calculated, we shall always first exclude the CPU cores listed in reserved-cpus from total capacity.
+This will result in the total CPU pool usable by Kubernetes. All other allocatable-decreasing, or safety-margin leaving feature shall take this adjusted capacity as its baseline when doing further calculations.
+An example of such an inter-working would be defining "kube-reserved", and "reserved-cpus" at the same time.
+In this case for example with "reserved-cpus" set to "1,9", and "kube-reserved" is set to "cpu=500m" on a 16 core system with hyper-threading disabled the allocatable capacity should be:
+16X1000m - len(1,9)X1000m - 500m = 14.500m.
```

Comment:
16X1000m - len(1,9)X1000m - 500m = 14.500m ? not 13500m ?

Reply:
Indeed it is!
I'm fin with superseding as well. however the later phases of the KEP is about making sure reserved-cpus is **always** taken into account. so, will it superseed kube-reserved and system-reserved when CPU-Manager is turned off, or policy is set to none?

BTW please note that the proposal was published a long time ago, and I stopped updating it because "it is not the immediate focus of SIG Node now". At least I remember there were some comments like that from Derek, though not sure where these disappeared to (maybe that discussion was on another thread?).
KEP is still open because I'm still holding out hope other companies will also discover the need of physically isolating infrastructure and workload CPUs from each other in all cases, on all nodes :)